Add a cache for metric metadata.

unbrice opened the pull request:
The DiskCache is implemented with lmdb, an on-disk file DB that can be accessed by multiple processes. Keys are metric names, values are json-serialised metadata.
In deployment the graphite storage dir is used as a rendez-vous point where all processes (carbon, graphite, ...) can find the metadata.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/criteo/biggraphite/20)
<!-- Reviewable:end -->

iksaif commented:
Reviewed 1 of 10 files at r4, 2 of 10 files at r7.
Review status: 3 of 11 files reviewed at latest revision, 5 unresolved discussions.

---

*[biggraphite/accessor.py, line 458 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5DXAYw_6BE3bv6_y_:-KJ5DXAZRanXrQxjiZmD:1646301424) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/accessor.py#L458)):*
> ```Python
>         # Finally, create the metric
>         metric_metadata_dict = metric_metadata.as_string_dict()
>         del metric_metadata_dict["name"]
> ```

why do we drop the name ?

---

*[biggraphite/metadata_cache.py, line 82 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5EIHSqyyOwnOq-XuR:-KJ5EIHSqyyOwnOq-XuS:1097150717) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/metadata_cache.py#L82)):*
> ```Python
>         if max_fds == resource.RLIM_INFINITY:
>             max_fds = self._MAX_READERS
>         max_fds = min(max_fds, self._MAX_READERS)
> ```

shouldn't this be significantly lower than RLIMIT_NOFILE ? (because fds are used for other things)

---

*[biggraphite/metadata_cache.py, line 98 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5EhEXZvimRFD6uVPu:-KJ5EhEXZvimRFD6uVPv:-798049972) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/metadata_cache.py#L98)):*
> ```Python
>             max_dbs=8,
>             # Should match max number of threads
>             max_spare_txns=128,
> ```

do we need that many threads ? (remember that we are likely to have multiple cache per machines anyway)

---

*[biggraphite/plugins/carbon.py, line 54 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5HbfDd8eAkxA6AcRe:-KJ5HbfEpodMeqlawvhA:-1420897928) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/plugins/carbon.py#L54)):*
> ```Python
>         except graphite_utils.ConfigException as e:
>             raise carbon_exceptions.CarbonConfigException(e)
>         storage_path = graphite_utils.storage_path_from_settings(settings)
> ```

should we make the cache optional ? (will be easier to measure performances with and without)

---

*[biggraphite/plugins/carbon.py, line 80 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5HhQaaQWswpBCmiDL:-KJ5HhQaaQWswpBCmiDM:10816313) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/plugins/carbon.py#L80)):*
> ```Python
>             carbon_xfilesfactor=xfilesfactor,
>         )
>         self._cache.create_metric(metadata)
> ```

do we need to handle the case where the metric is created locally and then disapear in the database (after one week for example). If not please explain why, if yes, please add a TODO, or a TTL in the cache.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/criteo/biggraphite/20)*
<!-- Sent from Reviewable.io -->

unbrice commented:
Review status: 1 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[biggraphite/accessor.py, line 458 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5DXAYw_6BE3bv6_y_:-KJ5JkRPJ9-hmibYY-qd:-1791826754) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/accessor.py#L458)):*
<details><summary><i>Previously, iksaif (Corentin Chary) wrote…</i></summary>
> why do we drop the name ?
</details>
it would be redundant with the first column, which is already the name. I added a comment.

---

*[biggraphite/metadata_cache.py, line 82 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5EIHSqyyOwnOq-XuR:-KJ5MNTWBC_FuGT1YK1N:143828214) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/metadata_cache.py#L82)):*
<details><summary><i>Previously, iksaif (Corentin Chary) wrote…</i></summary>
> shouldn't this be significantly lower than RLIMIT_NOFILE ? (because fds are used for other things)
</details>
Added doc, PTAL

---

*[biggraphite/metadata_cache.py, line 98 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5EhEXZvimRFD6uVPu:-KJ5b5MDoFOKYNCeqhzG:932383963) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/metadata_cache.py#L98)):*
<details><summary><i>Previously, iksaif (Corentin Chary) wrote…</i></summary>
> do we need that many threads ? (remember that we are likely to have multiple cache per machines anyway)
</details>
I improved the comment, PTAL

---

*[biggraphite/plugins/carbon.py, line 54 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5HbfDd8eAkxA6AcRe:-KJ5bNHsYzu5XmdgKuwn:-650975558) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/plugins/carbon.py#L54)):*
<details><summary><i>Previously, iksaif (Corentin Chary) wrote…</i></summary>
> should we make the cache optional ? (will be easier to measure performances with and without)
</details>
The cache follows the same interface as the accessor. During bench I'll hack the file to use the accessor as the cache. I don't think disabling the cache will ever be a good idea so I'm reluctant to making it configurable.

---

*[biggraphite/plugins/carbon.py, line 80 \[r7\]](https://reviewable.io:443/reviews/criteo/biggraphite/20#-KJ5HhQaaQWswpBCmiDL:-KJ5bfquJ5O2ZlDTd_VV:-112817696) ([raw file](https://github.com/criteo/biggraphite/blob/f806ddec474d819abc1c07b4099418463fdf0adf/biggraphite/plugins/carbon.py#L80)):*
<details><summary><i>Previously, iksaif (Corentin Chary) wrote…</i></summary>
> do we need to handle the case where the metric is created locally and then disapear in the database (after one week for example). If not please explain why, if yes, please add a TODO, or a TTL in the cache.
</details>
added a todo in metadata_cache.py

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/criteo/biggraphite/20)*
<!-- Sent from Reviewable.io -->

iksaif commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, coveralls wrote…</i></summary>
> 
> [![Coverage Status](https://coveralls.io/builds/6397709/badge)](https://coveralls.io/builds/6397709)
> 
> Coverage decreased (-0.2%) to 93.092% when pulling **9b6f2e06673c8266ac256422786e7293a64b1122 on unbrice:master** into **c04145592c766e623b2785d9b289e546be85bd8c on criteo:master**.
> 
</details>


---

Reviewed 2 of 2 files at r9.
Review status: 3 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/criteo/biggraphite/20#-:-KJ5hXMb6zlKORKgYPHt:1417088721)*
<!-- Sent from Reviewable.io -->
